src/core2/graph.js

```diff
@@ -250,60 +253,83 @@ export class Graph {
-    if (edge.address == null) {
-      throw new Error(`address is ${String(edge.address)}`);
+    const address = edge.address();
+    if (address == null) {
+      throw new Error(`address is ${String(address)}`);
     }
-    if (edge.src == null) {
-      throw new Error(`src is ${String(edge.src)}`);
+    const src = edge.src();
+    if (src == null) {
+      throw new Error(`src is ${String(src)}`);
     }
-    if (edge.dst == null) {
-      throw new Error(`dst is ${String(edge.dst)}`);
+    const dst = edge.dst();
+    if (dst == null) {
+      throw new Error(`dst is ${String(dst)}`);
     }
-    const srcIndex = this._addNodeAddress(edge.src);
-    const dstIndex = this._addNodeAddress(edge.dst);
-    const indexedEdge = {
-      address: edge.address,
-      srcIndex,
-      dstIndex,
-      payload: edge.payload,
-    };
-    return this._addIndexedEdge(indexedEdge);
-  }
-
-  _addIndexedEdge(indexedEdge: IndexedEdge): this {
-    const existingIndexedEdge = this._edges.get(indexedEdge.address);
-    if (existingIndexedEdge !== undefined) {
-      if (deepEqual(existingIndexedEdge, indexedEdge)) {
-        return this;
-      } else {
+    const existingEdge = this._edges.get(address);
+    if (existingEdge != null) {
+      if (!deepEqual(existingEdge.edge.toJSON(), edge.toJSON())) {
         throw new Error(
-          `edge at address ${JSON.stringify(
-            indexedEdge.address
-          )} exists with distinct contents`
+          `Edge at address ${stringify(address)} exists with distinct contents`
         );
       }
+      const srcAddress = src.address();
+      const existingSrcAddress = existingEdge.edge.src().address();
+      if (!deepEqual(srcAddress, existingSrcAddress)) {
+        throw new Error(
+          `Edge at address ${stringify(address)} exists with distinct src`
+        );
+      }
+      const dstAddress = dst.address();
+      const existingDstAddress = existingEdge.edge.dst().address();
+      if (!deepEqual(dstAddress, existingDstAddress)) {
+        throw new Error(
+          `Edge at address ${stringify(address)} exists with distinct dst`
+        );
+      }
+
+      return this;
+    }
+    // It's possible we were passed refs from a different graph. We will create a new edge
+    // with internal refs for consistency
+    if (src.graph() !== this || dst.graph() !== this) {
+      const edgeConstructor = findHandler(this._pluginMap, address.owner.plugin)
+        .createEdge;
+      edge = edgeConstructor(
+        this.ref(src.address()),
+        this.ref(dst.address()),
+        edge.toJSON()
+      );
     }
-    this._edges.add(indexedEdge);
-    this._outEdges[indexedEdge.srcIndex].push(indexedEdge.address);
-    this._inEdges[indexedEdge.dstIndex].push(indexedEdge.address);
+    // We make a redundant trip through the addressing layer; since
+    // src and dst are both refs, we've likely already done this.
+    // Potential perf optimization to internally expose the index directly.
```

Comment:
After discussion—I was misunderstanding the problem here. The above code
is sound, but wouldn’t do what we want, because almost always the
reference that we get is a delegate reference, so it won’t actually be
an instance of `IndexedReference`.

Furthermore, it is not unreasonable for a delegate reference to be
wrapping another delegate reference, so it’s not quite sufficient to
just say “if it’s a `DelegateReference`, unpack the internal base
reference, which should probably be an `IndexedReference`”—we’d have to
do this recursively, and even then this would only work with the
`DelegateReference` implementation that we know about, as opposed to any
other kind of implementation that users are totally free to use.

This is kind of a big problem, in my opinion: it changes the perf TODOs
from “not going to bother doing this right now” to “it’s not clear that
there’s actually a good way to do this, and we are incurring away what
might reasonably be a significant performance hit”.